Introduce Mailchimp Shortcode

artpi opened the pull request:
This PR introduces 2 things:
1. Mailchimp shortcode
2. Mailchimp gutenberg block.

More details in the master thread: p5uIfZ-8BV-p2

**The code behind this is already merged for WPCOM simple. If you want to test the UX - any simple site will work for proxied a12s**

## How this works

This enables visitors to sign up for a MailChimp list.

User needs to establish a Mailchimp connection in sharing section in calypso. There, they can use oauth flow to establish the connection and then select a list where subscribers will be saved.
( We imagine that this feature may be gated behing a plan - more info in the master thread )



### Shortcode

Once connection is established, shortcode `[jetpack-email-subscribe]` will display this signup form.

![screen shot 2018-11-12 at 18 14 01 pm](https://user-images.githubusercontent.com/3775068/48363970-76b06a00-e6a7-11e8-867b-e2be6927b1bf.png)

Once visitor signs up, her email is saved to a previously selected emaiil list in Mailchimp. We don't use it ourselves.
All email comes from the site owner through mailchimp

### Block

Once connection is established, site owner can use `jetpack/gutenberg` block.
If connection is NOT established, they get a prompt with the link to go to /sharing.


## Screenshots

### Block

![](https://user-images.githubusercontent.com/3775068/48910903-78c9b400-ee71-11e8-8a9d-ac85d5b11c87.png)

### Block, connection not set up


![zrzut ekranu 2018-12-28 o 15 30 05](https://user-images.githubusercontent.com/3775068/50518375-9e239080-0ab5-11e9-8639-ff887776c332.png)


## Testing instructions

1. Check out this PR on your Jetpack
2. Build blocks from calypso, the UI code is already merged
1. You need to log in as an Automattician
2. In /sharing you need to set up mailchimp connection and choose a list:
![screen shot 2018-11-12 at 18 13 20 pm](https://user-images.githubusercontent.com/3775068/48364106-d9096a80-e6a7-11e8-8513-5ea9279fd7e9.png)
3. Drop a `[jetpack-email-subscribe]` shortcode anywhere to test shortcode
4. Use `mailchimp` block for testing block.
5. Once you subscribe, your email will end up in the list set up in point 2.

artpi committed:
Port code from WPCOM

artpi committed:
Port code from WPCOM

artpi committed:
fix lint

artpi committed:
fix lint

enejb commented on modules/email-subscribe/email-subscribe.php:
```diff
@@ -0,0 +1,110 @@
+ *
+ * @hide-in-jetpack
+ */
+class Email_Subscribe {
```
We should prefix this class with Jetpack.

jeherve reviewed (changes requested):
Since you are adding a new shortcode, I would recommmend that you place it with the other shortcodes in Jetpack, as part of the Shortcodes module in `modules/shortcodes`.

jeherve commented on modules/email-subscribe/email-subscribe.php:
```diff
@@ -0,0 +1,110 @@
+	// Classic singleton pattern:
+	private static $instance;
+	private function __construct() {}
+	static function getInstance() {
```
Since the shortcode relies on the Subscriptions module, I would recommend that you check that Jetpack is active and that that the Subscriptions module is active before you do anything.

artpi replied:
It does not depend on subscription module anymore - we yanked the connection with subscriptions because of GDPR concerns. Now its a standalone thing.

jeherve commented on modules/email-subscribe/email-subscribe.php:
```diff
@@ -0,0 +1,110 @@
+
+	private function register_scripts_and_styles() {
+		wp_register_script( 'jetpack-email-subscribe', plugins_url( '/jetpack-email-subscribe.js', __FILE__ ), array( 'jquery' ) );
+		wp_register_style( 'jetpack-email-subscribe', plugins_url( '/jetpack-email-subscribe.css', __FILE__ ) );
```
I would suggest relying on `Jetpack::get_file_url_for_environment` here.

artpi replied:
Done!

jeherve commented on modules/email-subscribe/email-subscribe.php:
```diff
@@ -0,0 +1,110 @@
+		// We allow for overriding the presentation labels
+		$data = shortcode_atts( array(
+			'blog_id'     => $this->get_blog_id(),
+			'title'       => __( 'Join my email list' ),
```
You will want to add the `jetpack` textdomain to all those strings.

artpi replied:
Done!

jeherve reviewed (changes requested):
A few more remarks about things that could be improved.

artpi committed:
review feedback

artpi committed:
review feedback

artpi committed:
fix some paths

artpi committed:
fix some paths

jeherve commented on modules/shortcodes/email-subscribe.php:
```diff
@@ -0,0 +1,115 @@
+ *
+ * @hide-in-jetpack
+ */
+class Jetpack_Email_Subscribe {
```
I wonder if it would be worth describing things a bit here; what does the shortcode do exactly? Where do the users get subscribed? This way when we get back to this in 6 months we'll understand what this is used for.

artpi replied:
Description added.

jeherve commented on modules/shortcodes/email-subscribe.php:
```diff
@@ -0,0 +1,115 @@
+			wp_enqueue_script( 'jetpack-email-subscribe' );
+		}
+
+		if( ! wp_style_is( 'jetpack-email-subscribe', 'enqueue' ) ) {
```
You have a missing space here. Could you try to enable linting in your editor? That should help you catch the different coding standards issues (indenting, spaces) in that file.

artpi replied:
🤦‍♂️  OMG, funny story:
Since I work on payments stuff, I had to configure special install of PHPCS because payments standards are much more strict.
But PHPStorm can hold only 1 configuration in settings and the one from payments dies silently when used in other repos.
Which reports 0 errors to PHPStorm, so it looks like the file is clear.

You are right. That file was a mess. Fixed now.

artpi commented:
@jeherve I implemented all changes here :) If anything else needs improvement - please let me know :)

jeherve commented on modules/shortcodes/js/jetpack-email-subscribe.js:
```diff
@@ -0,0 +1,48 @@
+
+/* global jQuery */
+/* jshint esversion: 6, es3:false */
+const JetpackEmailSubscribe = {
```
It may be best to avoid the use of `const` here and everywhere else in the file. We try to stick to ES5 in other JS files that do not get compiled in Jetpack.

artpi replied:
YES! I would love that. I wrote that initially using `var`, but I got commit hook saying I should use `const`:

```
artpi~/GIT/jetpack(mailchimp/introduce-shortcode🔥 ) 😃 🚀  git add modules/shortcodes/js/jetpack-email-subscribe.js
artpi~/GIT/jetpack(mailchimp/introduce-shortcode) 😃 🚀  git commit -m "get rid of const"
Found '/Users/artpi/GIT/jetpack/.nvmrc' with version <10>
Now using node v10.13.0 (npm v6.4.1)
Found '/Users/artpi/GIT/jetpack/.nvmrc' with version <10>
Now using node v10.13.0 (npm v6.4.1)
husky > npm run -s precommit (node v10.13.0)


/Users/artpi/GIT/jetpack/modules/shortcodes/js/jetpack-email-subscribe.js
  11:1  error  Unexpected var, use let or const instead  no-var
  13:3  error  Unexpected var, use let or const instead  no-var
  17:3  error  Unexpected var, use let or const instead  no-var
  20:4  error  Unexpected var, use let or const instead  no-var
  22:4  error  Unexpected var, use let or const instead  no-var

✖ 5 problems (5 errors, 0 warnings)
COMMIT ABORTED: The linter reported some problems. If you are aware of them and it is OK, repeat the commit command with --no-verify to avoid this check.

husky > pre-commit hook failed (add --no-verify to bypass)

```

Now - is that my misconfiguration or the rules have changed?

artpi replied:
Done!

jeherve reviewed (changes requested):
A few more changes I would recommend.

artpi committed:
Coding standards

artpi committed:
Coding standards

artpi committed:
More standards fixes

artpi committed:
More standards fixes

artpi committed:
Revert const to var

artpi committed:
Revert const to var

artpi committed:
proper comment

artpi committed:
proper comment

simison commented on modules/shortcodes/email-subscribe.php:
```diff
@@ -0,0 +1,139 @@
+ */
+class Jetpack_Email_Subscribe {
+
+	private static $shortcode = 'jetpack-email-subscribe';
```
Should the shortcode be more Mailchimp specific? E.g. `jetpack-mailchimp-subscribe`? This seems pretty generic — unless it's in purpose and the actual backend service could be switched over to anything else in future?

artpi replied:
>  it's in purpose and the actual backend service could be switched over to anything else in future?

That is precisely the case.

simison commented on modules/shortcodes/email-subscribe.php:
```diff
@@ -0,0 +1,139 @@
+				'consent_text'      => __( 'By clicking submit, you agree to share your email address with the site owner and MailChimp to receive marketing, updates, and other emails from the site owner. Use the unsubscribe link in those emails to opt out at any time.', 'jetpack' ),
+				'processing_label'  => __( 'Processing...', 'jetpack' ),
+				'success_label'     => __( 'Success! You\'ve been added to the list.', 'jetpack' ),
+				'error_label'       => __( "Oh no! Unfortunately there was an error.\nPlease try reloading this page and adding your email once more.", 'jetpack' ),
```
I don't remember seeing `\'` or `\n` in our translation strings previously, is that okay?

artpi replied:
I don't know. We need multiline here, so I guess its better than `br` ?

artpi committed:
Remove email type and group selectors

artpi committed:
Remove email type and group selectors

artpi committed:
drop placeholder styling entirely

artpi committed:
drop placeholder styling entirely

artpi commented on modules/shortcodes/css/jetpack-email-subscribe.css:
```diff
@@ -0,0 +1,59 @@
+}
+input[type=email].jetpack-email-subscribe-email:-moz-placeholder {
+    color: #ddd;
+}
```
I got rid of the placeholder styling entirely since I cannot make it work.

jeherve commented on modules/shortcodes/email-subscribe.php:
```diff
@@ -0,0 +1,139 @@
+	 */
+	public static function get_instance() {
+		// Do not load this at all if it's neither a WPCOM or a connected JP site.
+		if ( ! ( ( defined( 'IS_WPCOM' ) && IS_WPCOM ) || Jetpack::is_active() ) ) {
```
I would check if the Subscriptions module is added as well, since you rely on it in the shortcode.

artpi replied:
I am not relying on it. It *does not* require subscriptions module to be active to work :)

jeherve replied:
Ok then. Do you think you could update the testing instructions so we can now what to expect when signing up with that shortcode: where is your email address sent when subscribing there, what emails should be sent to you after you've subscribed?

artpi replied:
Updated!

jeherve reviewed (changes requested):
One more quick comment

artpi commented:
Are there any other changes I should do here?

artpi committed:
Add Gutenberg block

jeherve reviewed (changes requested):
I gave it a try, it works well. I would have a few more remarks:

- When I successfully subscribe, Mailchimp shows the subscriber as coming from `WordPress.com`. Is there any way to change that to be more specific?

- Would there be a way to return a different error message depending on the kind of error we got when trying to subscribe? As of right now you only get one error message, that may not always be helpful: was I already subscribed with that email address, was there a communication issue between wpcom and my site, between wpcom and Mailchimp?

artpi commented on modules/shortcodes/js/jetpack-email-subscribe.js:
```diff
@@ -0,0 +1,48 @@
+			emailField.removeClass( cssPrefix + '-form-error' );
+			var email = emailField.val();
+
+			if ( ! JetpackEmailSubscribe.validateEmail( email ) ) {
```
Yes, this is a valid criticism and stems from the fact, that we plan to:
- roll out with MailChimp
- we will be looking into other solutions pretty soon
- we want to provide an infrastructure for easy replacing the actual provider without any change needed on the Jetpack site itself OR the shortcode on the sites. If the shortcode was named "mailchimp", but suddenly we would provide back-end drop-in replacement with a different service, that would be very confusing.

So we have to keep the skeleton as "email subscription" with a clear intent of setting up mailchimp (for now) and later we have to make clear to the site owner which service they currently have connected and that they can change it without re-editing posts.

artpi commented on modules/shortcodes/js/jetpack-email-subscribe.js:
```diff
@@ -0,0 +1,48 @@
+/*
```
I am not sure I understand the simplicity argument, since

a) the code is minified, so the filesize is not an issue
b) using jQuery is explicit and there is no confusion for what it stands for. After longer work with vanilla JavaScript in chrome, $ starts being remembered last-viewed-dom-element and not jquery
c) It is clear at every point that we are not going to fall into jQuery conflict trap, since we are not using shorthand

jeherve replied:
👍 	That is really minor so more nit-picking than anything else. We tend to alias jQuery in other Jetpack files but that is not a must.

artpi committed:
Make some attributes non-editable

artpi committed:
Make some attributes non-editable

artpi commented on modules/shortcodes/email-subscribe.php:
```diff
@@ -0,0 +1,139 @@
+		// We allow for overriding the presentation labels.
+		$data = shortcode_atts(
+			array(
+				'blog_id'           => $this->get_blog_id(),
```
I made it non-editable

mdawaffe replied:
It looks like we lost the changes from https://github.com/Automattic/jetpack/pull/10597/commits/46b63a30bc852431eea1fc075a00a9e5365c9aca?

artpi commented on modules/shortcodes/email-subscribe.php:
```diff
@@ -0,0 +1,139 @@
+				'processing_label'  => __( 'Processing...', 'jetpack' ),
+				'success_label'     => __( 'Success! You\'ve been added to the list.', 'jetpack' ),
+				'error_label'       => __( "Oh no! Unfortunately there was an error.\nPlease try reloading this page and adding your email once more.", 'jetpack' ),
+				'classname'         => self::$css_classname_prefix,
```
I made it non-editable

artpi commented:
Thanks for testing!
I responded to comments

> When I successfully subscribe, Mailchimp shows the subscriber as coming from WordPress.com. Is there any way to change that to be more specific?

This is the name of the application that is registered with MailChimp Oauth. We can rename it to anything else, but it will be consistent among every user that is our customer.

> Would there be a way to return a different error message depending on the kind of error we got when trying to subscribe? As of right now you only get one error message, that may not always be helpful: was I already subscribed with that email address, was there a communication issue between wpcom and my site, between wpcom and Mailchimp?

Yes, I have those error codes to provide better feedback in the future. The problem is that some error messages you want to keep to yourself (this is spam!), other ones you want to show to the site owner only (your mailchimp oauth token is wrong ) and yet another you want to show to the user directly.
This introduces complexity in design, engineering and copywriting that we definitely don't have resources to deal with in V1. This was a conscious decision to put it off for later, but the work has already started.

jeherve commented:
Thanks for the replies!

artpi committed:
make it work for sandbox

artpi committed:
code from sandbox

artpi committed:
Merge branch 'mailchimp/gutenberg' into mailchimp/introduce-shortcode

artpi committed:
enable

artpi committed:
Add fix for the rest route

artpi committed:
fix for old php

artpi commented:
I made following changes here:
a) Block itself was rebranded as "mailchimp" after deliberatinons with @lessbloat . We will probably introduce another block if we provide more 
b) I merged block into this PR so they can go together
c) I provided a "not set up yet" screen
d) Calypso and wpcom pieces are merged
e) This has to go out in the next release.

artpi committed:
Add a filter to deal with SSR problem

jeherve reviewed (changes requested):
I tested the block and would have some feedback; 

- Should the block title be "Mailchimp form", to match the "Subscription form" block?
- Can the different fields be edited in the editor instead of only in the inspector controls? This would be more consistent with all our other blocks.
- It would be nice if the block was saved as a shortcode behind the scenes, with the different parameters saved as shortcode parameters. I believe that is how the other blocks work.
- I first typed `/subscr`, hoping to get the block to popup, but it did not. Maybe it should be one of the keywords?

Both the shortcode and the block rely on the API call to subscribe, and I found myself missing some information when attempting to sign in with an address that was already subscribed to the site. I received the standard error message in the form, but the API call had actually returned the following error: `{"error":"member_exists"}` Maybe we should surface those errors somehow, have them appear after the custom message?

When I originally connected my site to Mailchimp, it wasn't clear to me that I needed to go back to my site and refresh the editor. It may be good to mention it somewhere in the flow maybe?

I believe you will need to add your block name to this list:
https://github.com/Automattic/jetpack/blob/0da5e5d398188875b855b01a8c668f8fcc29e1a7/class.jetpack-gutenberg.php#L65
cc @ockham for confirmation: looking at that list right now I see blocks that are in beta (subscriptions), and I don't see blocks like the Tiled Gallery one?

artpi commented on modules/shortcodes/email-subscribe.php:
```diff
@@ -0,0 +1,233 @@
+	private function register_init_hook() {
+		add_action( 'init', array( $this, 'init_hook_action' ) );
+		add_action( 'jetpack_options_whitelist', array( $this, 'filter_whitelisted_options' ), 10, 1 );
+		add_action( 'jetpack_blocks_to_register', array( $this, 'prevent_jetpack_register_block' ), 10, 1 );
```
Response here https://github.com/Automattic/jetpack/pull/10597#discussion_r245018740

ockham commented:
> I believe you will need to add your block name to this list:
> https://github.com/Automattic/jetpack/blob/0da5e5d398188875b855b01a8c668f8fcc29e1a7/class.jetpack-gutenberg.php#L65
> cc @ockham for confirmation: looking at that list right now I see blocks that are in beta (subscriptions), and I don't see blocks like the Tiled Gallery one?

Yeah, this is confusing, sorry. Those `$default_blocks` are a fallback -- Jetpack will look for and `index.json` (listing blocks both in production and beta) in the blocks directory first, which it gets[ from Calypso](https://github.com/Automattic/wp-calypso/blob/f890d857f7d9dbef6646049981c2e5133c23e717/client/gutenberg/extensions/presets/jetpack/index.json)/the SDK as part of the build process. Since `mailchimp` is included there already, it doesn't need to be listed as a `$default_block`.

(I'm planning to remove that source of confusion soon.)

artpi commented on modules/shortcodes/email-subscribe.php:
```diff
@@ -0,0 +1,233 @@
+
+	private function register_gutenberg_block() {
+		if ( Jetpack_Gutenberg::is_gutenberg_available() ) {
+			register_block_type( 'jetpack/' . self::$block_name, array(
```
There is a bug in Jetpack_Gutenberg  which makes Server-side rendered blocks impossible in Jetpack.

- `jetpack_register_block` makes `register_block_type` call too late for the route to SSR blocks to register. That is too late to WP_REST_Block_Renderer_Controller be aware of them.
- Calling `register_block_type` does make it register, but does not show the SDK blocks.
- calling both produces a notice in the logs.

Tne ONLY way of getting out of this conundrum is adding this filter OR registering a route for the SSR path seperately. I tried it and it broke MC, because MC requires registering routes in a special hook which makes no sense for Jetpack.

This is a catch-22. This filter is meant as a way of injecting jetpack_register_block

ockham replied:
> There is a bug in Jetpack_Gutenberg which makes Server-side rendered blocks impossible in Jetpack.

I think we have some instances of working server-side rendered blocks, e.g. [Contact Form](https://github.com/Automattic/jetpack/blob/master/modules/contact-form/grunion-contact-form.php).

This issue sounds familiar. I believe that we've run into it before, and I think we have a fix (or at least an established workaround) available. I'll need to study the code some more before being able to provide a better informed opinion.

artpi replied:
If there is a more elegant fix, I am happy to take it!

ockham replied:
So I just tried this

```diff
diff --git a/modules/shortcodes/email-subscribe.php b/modules/shortcodes/email-subscribe.php
index bc2564832..867059901 100644
--- a/modules/shortcodes/email-subscribe.php
+++ b/modules/shortcodes/email-subscribe.php
@@ -57,7 +57,7 @@ class Jetpack_Email_Subscribe {
 	private function register_init_hook() {
 		add_action( 'init', array( $this, 'init_hook_action' ) );
 		add_action( 'jetpack_options_whitelist', array( $this, 'filter_whitelisted_options' ), 10, 1 );
-		add_action( 'jetpack_blocks_to_register', array( $this, 'prevent_jetpack_register_block' ), 10, 1 );
+		//add_action( 'jetpack_blocks_to_register', array( $this, 'prevent_jetpack_register_block' ), 10, 1 );
 	}
 
 	public function prevent_jetpack_register_block( $blocks ) {
@@ -79,7 +79,7 @@ class Jetpack_Email_Subscribe {
 
 	private function register_gutenberg_block() {
 		if ( Jetpack_Gutenberg::is_gutenberg_available() ) {
-			register_block_type( 'jetpack/' . self::$block_name, array(
+			jetpack_register_block( self::$block_name, array(
 				'attributes' => array(
 					'title' => array(
 						'type' => 'string',
@@ -108,8 +108,7 @@ class Jetpack_Email_Subscribe {
 				),
 				'style' => 'jetpack-email-subscribe',
 				'render_callback' => array( $this, 'parse_shortcode' ),
-			) );
-			jetpack_register_block( self::$block_name, array(), array( 'available' => true ) );
+			), array( 'available' => true ) );
 		}
 	}
 
```

and the block seems to work both in the editor (and is availalble in the block picker), and in the frontend. Am I missing anything else that this would break?

artpi replied:
Yes. What breaks is Gutenberg on WPCOM. If you make that change in WPCOM codebase, this is what you get in Gutenlypso:

<img width="927" alt="zrzut ekranu 2019-01-03 o 17 28 08" src="https://user-images.githubusercontent.com/3775068/50649038-10e0b000-0f7d-11e9-9220-f4ea71c786af.png">

And since files have to be kept in sync, any fix has to be here too.

ockham replied:
Okay, I'll have a closer look at the WP.com side of things then :+1:

artpi replied:
Thank you! :)

mdawaffe replied:
If we don't have a better solution for this, can this be documented in code comments? The combination of `prevent_jetpack_register_block()`, `register_gutenberg_block()`, and `jetpack_register_block()` is pretty confusing.

ockham replied:
Haven't been able to find a more generic fix yet (planning to get some other things out of the way first to make this easier to reason about, see pafL3P-cY-p2).

But yeah, definitely agree that this should be well-documented in code so we can cycle back and remove once we have a better solution.

ockham replied:
I'm confident that #11091 will fix this (and allow us to remove the workarounds that this PR introduced).

jeffersonrabb commented:
> Once connection is established, shortcode `[jetpack-email-signup]` will display this signup form.

I think this is a typo, and the shortcode is actually `[jetpack-email-subscribe]` (as it appears later in the PR description)

artpi committed:
Rename to Mailchimp form

artpi commented on modules/shortcodes/email-subscribe.php:
```diff
@@ -0,0 +1,233 @@
+
+	private static $css_classname_prefix = 'jetpack-email-subscribe';
+
+	private static $option_name = 'jetpack_mailchimp';
```
This is generally very debateable.

[We plan to add different "email subscription engines"](https://github.com/Automattic/jetpack/pull/10597#discussion_r236598803), so the shortcode could be made flexible, since that is the pattern with shortcodes. We dont want to have too many of them, because people have to remember about them. 

The case with blocks is exactly the opposite - they serve as a discovery method, so we plan to make seperate blocks for each service, powered by the code behind the shortcode.

enejb replied:
cc: @mapk 

I think we should be careful with introducing too many different blocks here. How is this block different from our current subscription block? Instead of adding another "Mailchimp" subscription block why not add options or setting that turn the current subscription block into the one that is powered by Mailchimp?

artpi commented:
> Should the block title be "Mailchimp form", to match the "Subscription form" block?

Changed here, will change in calypso

> Can the different fields be edited in the editor instead of only in the inspector controls? This would be more consistent with all our other blocks.

Since this is server-side-rendered block for the purposes of MVP, this will have to be improved later

> It would be nice if the block was saved as a shortcode behind the scenes, with the different parameters saved as shortcode parameters. I believe that is how the other blocks work.

Agree 100%. [This is how I started this whole block, but feedback has led me to abandon this line, since it would require me to amend Jetpack React dependencies and I didn't want to deal with that.
This can be improved in another PR.](https://github.com/Automattic/wp-calypso/pull/28798#pullrequestreview-184628690)

> I first typed /subscr, hoping to get the block to popup, but it did not. Maybe it should be one of the keywords?

Good point, happy to do it in Calypso.

> Both the shortcode and the block rely on the API call to subscribe, and I found myself missing some information when attempting to sign in with an address that was already subscribed to the site. I received the standard error message in the form, but the API call had actually returned the following error: {"error":"member_exists"} Maybe we should surface those errors somehow, have them appear after the custom message?

All valid points, but this is meant as a V1.
Error messaging can be improved and there are few issues we can work out with that.

> When I originally connected my site to Mailchimp, it wasn't clear to me that I needed to go back to my site and refresh the editor. It may be good to mention it somewhere in the flow maybe?

Yes, we will improve that in the future.

> I believe you will need to add your block name to this list:

As @ockham mentioned - it works now without that

> I think this is a typo, and the shortcode is actually [jetpack-email-subscribe] (as it appears later in the PR description)

Yes. Will fix.

> This issue sounds familiar. I believe that we've run into it before, and I think we have a fix (or at least an established workaround) available. I'll need to study the code some more before being able to provide a better informed opinion.

If there is a more elegant fix, I am happy to take it!

artpi committed:
Add css warning

artpi committed:
remove badly added file

jeffersonrabb reviewed (dismissed):
Most of my comments have to do with the Block and Mailchimp connect UI, so I left them in the merged `wp-calypso` PR here: https://github.com/Automattic/wp-calypso/pull/28798#issuecomment-451243149

No issues with the shortcode, which worked well in a self-hosted Jetpack site and Calypso.

mdawaffe commented on modules/shortcodes/css/jetpack-email-subscribe.css:
```diff
@@ -0,0 +1,53 @@
+ * WARNING: This file is distributed verbatim in Jetpack.
+ * There should be nothing WordPress.com specific in this file.
+ *
+ * @hide-in-jetpack
```
We generally don't include these comments in Jetpack. Only in WordPress.com's codebase.

artpi replied:
Removed

mdawaffe commented on modules/shortcodes/email-subscribe.php:
```diff
@@ -0,0 +1,233 @@
+ * WARNING: This file is distributed verbatim in Jetpack.
+ * There should be nothing WordPress.com specific in this file.
+ *
+ * @hide-in-jetpack
```
We generally don't include these comments in Jetpack. Only in WordPress.com's codebase.

artpi replied:
Removed!

mdawaffe commented on modules/shortcodes/email-subscribe.php:
```diff
@@ -0,0 +1,233 @@
+
+/**
+ * Class Jetpack_Email_Subscribe
+ * This class encapsulates shortcode for subscribing to a MailChimp list.
```
```suggestion
 * This class encapsulates a shortcode for subscribing to a MailChimp list.
```

mdawaffe commented on modules/shortcodes/email-subscribe.php:
```diff
@@ -0,0 +1,233 @@
+/**
+ * Class Jetpack_Email_Subscribe
+ * This class encapsulates shortcode for subscribing to a MailChimp list.
+ * It displays a simple signup form that gets an email address from the user and signs him for a list
```
```suggestion
 * The shortcode displays a simple signup form. Submitting an email address will sign the visitor up
```

artpi replied:
Thank you!

artpi replied:
I added a different comment that mentions gutenberg.

mdawaffe commented on modules/shortcodes/email-subscribe.php:
```diff
@@ -0,0 +1,233 @@
+ * Class Jetpack_Email_Subscribe
+ * This class encapsulates shortcode for subscribing to a MailChimp list.
+ * It displays a simple signup form that gets an email address from the user and signs him for a list
+ * selected in "Sharing" section in calypso.
```
```suggestion
 * for the Mailchimp list selected in the "Sharing" section of Calypso.
```

mdawaffe commented on modules/shortcodes/js/jetpack-email-subscribe.js:
```diff
@@ -0,0 +1,48 @@
+ * WARNING: This file is distributed verbatim in Jetpack.
+ * There should be nothing WordPress.com specific in this file.
+ *
+ * @hide-in-jetpack
```
We generally don't include these comments in Jetpack. Only in WordPress.com's codebase.

artpi replied:
removed

mdawaffe commented on modules/shortcodes/js/jetpack-email-subscribe.js:
```diff
@@ -0,0 +1,48 @@
+		var re = /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
+		return re.test( String( email ).toLowerCase() );
+	},
+	activate: function( blogId, domId, cssPrefix ) {
```
If these parameters are editable by non-Editors/Admins, odd things can happen. I believe that was part of the point of https://github.com/Automattic/jetpack/commit/46b63a30bc852431eea1fc075a00a9e5365c9aca, but that commit seems to have been reverted.

artpi replied:
Amazing catch ! Thank you -  I rebased it badly.

mdawaffe commented on modules/shortcodes/js/jetpack-email-subscribe.js:
```diff
@@ -0,0 +1,48 @@
+			}
+
+			jQuery( '#' + domId + ' form' ).hide();
+			jQuery( '.' + cssPrefix + '-processing' ).css( 'display', 'block' );
```
What's the difference between `.show()` and `.css( 'display', 'block' )`?

artpi replied:
`show` add animation

artpi replied:
done

mdawaffe reviewed (changes requested):
Other than the points about 46b63a3's changes being missing, the comments I made are not blockers.

mdawaffe committed:
Update modules/shortcodes/email-subscribe.php

Co-Authored-By: artpi <artur.piszek@gmail.com>

mdawaffe committed:
Update modules/shortcodes/email-subscribe.php

Co-Authored-By: artpi <artur.piszek@gmail.com>

artpi committed:
Remove comments

artpi committed:
File block comment fixes

artpi committed:
Add comment about the fix

artpi commented:
> If we don't have a better solution for this, can this be documented in code comments? The combination of prevent_jetpack_register_block(), register_gutenberg_block(), and jetpack_register_block() is pretty confusing.

Thank you for catching that!
I swear, there was a comment that got lost in the rebase. Added one now.

artpi committed:
move to show

artpi committed:
fix with display

artpi committed:
If member is already on the list - show success

artpi committed:
Add title as an optional parameter

artpi committed:
reduce size

artpi commented:
I implemented following fixes from @jeffersonrabb comments:

> 2. After a failed email submission, the error message is generic and recommends a page refresh. These messages should indicate the reason for failure. I imagine the most common cause of failure will be an email that's already on the list, and if the user isn't informed of this they'll try over and over until they get frustrated.

Yes, this is true. When I think about it its not actually the error - end state should be the same as success. User is on the list.
I resolve this by showing success message.

> 7. This one could probably go either way, but there seems to be a Gutenberg convention of not building titles in blocks.

good point. I made title an optional parameter and don't include a default one, so its not showing.
However, for quick copy-pasting of shortcodes and quicker block setup, you can set it.

> 8. If there are multiple instances of the block (or shortcode) on a page, clicking Submit in one appears to submit them all. The submitted button state is applied to all, and all get the same success/failure notification.

Yes. I fixed that.

artpi commented:
I have included recent changes in this PR to D22833-code .
On monday, Ill merge it and rebase D22705-code 

I dont want to merge D22705-code  until we are sure that we want this fix for SSR to be merged.

jeherve reviewed (approved):
This looks good to me now. 👍

enejb commented on modules/shortcodes/email-subscribe.php:
```diff
@@ -0,0 +1,242 @@
+	 *
+	 * @return array
+	 */
+	public function prevent_jetpack_register_block( $blocks ) {
```
We are doing something wrong here. We should always register the block using jetpack_register_block 
If you have specific visibility requirement that depend blog info (options, plan etc) You will need to use the callback. See https://github.com/Automattic/jetpack/blob/master/modules/simple-payments/simple-payments.php#L61 as an example.

jeherve replied:
I believe this was discussed here:
https://github.com/Automattic/jetpack/pull/10597#discussion_r245018740

jeherve reviewed (approved):
Works well for me now. Merging.

jeherve commented:
Cherry-picked to `branch-6.9` in 2e30de4a902eeb880a70ef5d798c2f0427dec5c1

jeffersonrabb commented:
Proposed change to the block here: https://github.com/Automattic/wp-calypso/pull/29929. Removes Server-Side Rendering, could simplify some of the registration questions.
